src/ray/object_manager/pull_manager.cc

```diff
@@ -96,30 +98,37 @@ void PullManager::TryToMakeObjectLocal(const ObjectID &object_id) {
     return;
   }
 
+  // New object locations were found, so begin trying to pull from a
+  // client. This will be called every time a new client location
+  // appears.
+  bool did_pull = PullFromRandomLocation(object_id);
+  if (did_pull) {
+    UpdateRetryTimer(request);
+    return;
+  }
+
+  // If we cannot pull, try restoring objects from the external storage.
   if (!request.spilled_url.empty()) {
-    // Try to restore the spilled object.
+    const auto spilled_node_id = request.spilled_node_id;
+    RAY_CHECK(!spilled_node_id.IsNil());
+    // Try to restore the spilled object from a local or remote node.
     restore_spilled_object_(
-        object_id, request.spilled_url, [this, object_id](const ray::Status &status) {
-          bool did_pull = true;
-          // Fall back to fetching from another object manager.
+        object_id, spilled_node_id,
+        [object_id, spilled_node_id](const ray::Status &status) {
           if (!status.ok()) {
-            did_pull = PullFromRandomLocation(object_id);
-          }
-          if (!did_pull) {
-            RAY_LOG(WARNING) << "Object restoration failed and the object could not be "
-                                "found on any other nodes. Object id: "
-                             << object_id;
+            RAY_LOG(WARNING)
+                << "Object restoration request to a remote failed and the object could "
+                   "not be "
+                   "found on any other nodes. It could be because the remote node that "
+                   "spilled the object has crashed or the network request has failed. "
```

Comment:
```suggestion
                   "found on any other nodes. This can happen if the location where the object was spilled is unreachable. This job may hang if the object is permanently unreachable."
```